[otp_ctrl/doc] Initial commit of OTP spec

msfschaffner opened the pull request:
Note that this is a draft PR, and hence a WIP (I'm out for the rest of the week, but will finalize this at the beginning of next week).

I've marked the sections that need some more details with TODOs in the document (e.g., the Life Cycle interface, the ciphering and digest mechanism, a detailed block diagram, as well as links to docs that do not exist yet). Also, some of the content may have to be condensed down a bit.

Signed-off-by: Michael Schaffner <msf@opentitan.org>

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+The proprietary OTP IP provides:
+- Reliable, non-volatile storage.
+- Inclusive of any technology-specific redundancy or error correction mechanisms.
```
i think we can get rid of `inclusive of any`

msfschaffner replied:
yeah let's remove that.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+- Separate life cycle partition and interface to life cycle controller
+	- Supports life cycle functions, but cannot be integrity locked
+- Lightweight scrambling of secret OTP partition using a global netlist constant
+- Lightweight ephemeral key derivation function for RAM scrambling mechanisms
```
did you want to mention the flash key also? That one technically isn't ephemeral, but the OTP does still "churn" on it.

msfschaffner replied:
ah yes, good point. I've added a line for FLASH without the word "ephemeral".

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
 
-## Block Diagram
+Conceptually speaking, the OTP functionality is at a high level split into "front-end" and "back-end".
+The "front-end" contains the logical partitions and associated buffer registers that feed the hardware and software consumer interfaces of the system.
```
ah this is probably legacy from the old spec. It implies there that the buffered registers also feed software.  Probably should update that.

msfschaffner replied:
Yeah, I've actually updated the diagram to show that the "frontend" can have both buffered and unbuffered partitions.
I've removed the `and associated buffer registers` part of this sentence, since further below it is explained that some of the partitions may be buffered, while others may be unbuffered.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+- Secrecy
+  - This controls whether a particular partition contains secret data
+  - If secret, a partition is not readable by software once locked, and is encrypted in storage
```
@zi-v 
just FYI

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+Since the OTP is memory-like in nature (it only outputs a certain number of bits per address location), some of the logical partitions are buffered in registers for instantaneous and parallel access by hardware.
+This is a critical point, as after power-up, OTP contents are stored in flip flops and sourced to the system.
+They are **NOT** directly sourced from the OTP macro itself.
```
o this line is probably also legacy, should probably clarify software configuration bits are sourced from the OTP directly.

msfschaffner replied:
ah, yes let me rephrase to make it more clear. this line should only refer to the buffered partitions.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+Generally speaking, the production life cycle of a device is split into 5 stages "Manufacturing" -> "Calibration and Testing" -> "Provisioning" -> "Mission" -> "RMA".
+OTP values are usually programmed during "Calibration and Testing", "Provisioning" and "RMA" stages, as explained below.
+A detailed listing of all the collateral and the corresponding memory map can be found in the Programmer's Guide Section further below.
```
maybe it could be good to include a link to the `programmers guide` section

msfschaffner replied:
Yes I've tried that but without success... for some reason I did not manage to create a link that HUGO would render correctly...

msfschaffner replied:
ah now it works - forgot to strip away the apostrophe :D

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+The secret partition is populated with root secrets and keys that are critical to establishing the device identity.
+
+As part of injecting the final firmware, the SKU-specific hardware and software configurations are also programmed.
+
```
@moidx 
@zi-v 
@cdgori 
@gkelly 

it would be good if you guys can review these two sections `Calibration and Test` and `Provisioning` and make sure it is a reasonable representation.  This text here is just meant to be give the reader a sample, but does not need to be perfectly accurate.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+## Locking a Partition
+
+Write access to a partition can be permanently locked when software determines it will no longer make any updates to that partition.
+To lock, a truncated integrity constant (32bit) is calculated and programmed alongside the other data of that partition.
```
is the integrity constant still 32b in the new world? I THINK it is..but just in case it isn't, probably good to update this.

msfschaffner replied:
yes, the integrity constants are still truncated to 32bit. however, given that the digest is taken from the 64bit scrambler output, we could just align this with 64bit and not truncate at all... WDYT?

msfschaffner replied:
actually, let's just leave this at 32bit in order to align the HW digest sizes with the SW digests.

msfschaffner replied:
I have aligned the integrity constants with the size granule of the partition, meaning: 32bit for SW and HW config partitions, 64bit for secret partitions.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+### Software Configuration Partition
+
+The software configuration partition uses the OTP as non-volatile flags and storage.
```
probably should add "configuration" also (such as AST)

msfschaffner replied:
reworded to:
`The software configuration partitions are used as non-volatile storage for flags, configuration and calibration data.`

also, I made the `Partition**s**` plural in the title

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+The software configuration partition uses the OTP as non-volatile flags and storage.
+As such, the contents of this partition are usually consumed once as part of code execution, or moved to another storage compartment somewhere in the design (for example, the clock calibration values are programmed to the clocking module, and the LDO calibration values are programmed to the regulator wrapper).
+
```
oops the text above here is out of date.  I think both things I use as examples here are programmed to the analog wrappers.

msfschaffner replied:
ah yeah, good catch. I'll rephrase that example and mention that these get programmed into AST.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+These values are stored in OTP as plaintext.
+
+Secret partitions contain data that are critical to security, for example flash scrambling keys, device root secret and unlock tokens.
+These values are stored encrypted in OTP, and are decrypted upon read.
```
i guess we should change the terms `encrypted` to `scrambled` right?

msfschaffner replied:
oh yeah indeed - fixed!

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+
+## Power-up and Sense
+
+The OTP controller partition storage must output a specified default (it is not always 0 like a blank OTP) upon reset release.
```
@cdgori 
do you think it makes sense to elaborate more on the defaults here?
It should in theory be "safe" defaults, meaning things are "disabled" by default.

tjaychen replied:
o wait nvm, i guess the section below talks about this.

msfschaffner replied:
I see what you mean... I've added the word "safe" before "default" in this sentence

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+To lock, a truncated integrity constant (32bit) is calculated and programmed alongside the other data of that partition.
+
+Once the "integrity digest" is non-zero, no further updates are allowed.
+If the partition is secret, software is in addition no longer able to read the contents (see secret partiton description further below).
```
maybe add a soft link for "secret partition description" . `partiton` -> `partition`

msfschaffner replied:
good idea, will do

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+Instead, software can simply check as part of secure boot and take other measures when these values are programmed into peripherals.
+
+For this partition it is thus the responsibility of software to calculate the integrity digest and program it into the OTP.
+It is also reasonable to shadow (parts) of this partition in main memory, and there is not an immediate impact from OTP contents to hardware.
```
do you mean shadow (parts of) this partition in main memory

msfschaffner replied:
jep, thanks :)

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+### Hardware Configuration and Secret Partitions
+
+The hardware and secret partitions directly affect downstream hardware.
+The contents must go through periodic integrity checks and so the stored digest is calculated by hardware when software provides the intent to lock.
```
The contents must go through periodic integrity checks and (so-> redundant?) the stored digest is calculated by hardware when software provides the intent to lock.

msfschaffner replied:
hm... "so" should be a "hence" or a "therefore" in this case. I want to point out that the periodic hw checks are the reason this digest needs to be calculated in hw (as opposed to sw as for the sw partitions).

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+I.e., values programmed to OTP via the programming interface will not be visible in buffered registers until after the next power cycle.
+
+At this point, outputs of the partition storage are NOT expected to change unless a periodic check suddenly fails.
+When this failure occurs, all outputs are reverted to its default state, and an alert is immediately triggered to the alert manager.
```
shall we call it alert handler instead?

msfschaffner replied:
yes that would be more consistent. I'll change that.

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +296,418 @@ The following table lists the main parameters used throughout the OTP controller
+
+1. Write the Byte address for the access to {{< regref DIRECT_ACCESS_ADDRESS >}}.
+Note that the address is aligned with the granule, meaning that either 2 or 3 LSBs of the address are ignored, depending on whether the access granule is 32 or 64bit.
+2. Trigger a write command by writing 0x1 {{< regref DIRECT_ACCESS_CMD >}}.
```
Trigger a read command?

msfschaffner replied:
copy paste :)... fixed

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +296,418 @@ The following table lists the main parameters used throughout the OTP controller
+3. Trigger a write command by writing 0x2 {{< regref DIRECT_ACCESS_CMD >}}.
+4. Poll the {{< regref STATUS >}} until the state goes back to IDLE or ERROR.
+Alternatively, the `otp_operation_done` interrupt can be enabled up to notify the processor once an access has completed.
+5. If the operation has finished with ERROR status, additional handling is required (see Section on Error handling).
```
maybe add the soft link ofr Error handling? (also applies to line 562)

msfschaffner replied:
jup good idea - added!

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+  - This controls whether a partition is locked and prevented from future updates.
+  - A locked partition is stored alongside a 32b digest (truncated hash) to be used later for integrity verification.
+- Integrity Verification
+  - Once a partition is locked, it can undergo periodic verification (time-scale configurable by software)
```
When you mention once a partition is locked -> does this `locked` applies to both read lock and write lock, or it is write lock only?

msfschaffner replied:
Integrity verification applies to partitions that get write-locked after burning in a digest value.

Note that for read locks there are two mechanisms as described above. SW_CFG can be locked at runtime via a CSR. secret partitions get statically read locked whenever they get write locked (these read and write locks are coupled to the same digest.

msfschaffner replied:
I'll amend that sentence to make it clear.

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +296,418 @@ The following table lists the main parameters used throughout the OTP controller
+
+![OTP Controller Block Diagram](otp_controller_block_diagram.svg)
+
+Note the block diagram shows two PRESENT ciphers, although in reality they should be shared between the functions.
```
two PRESENT ciphers? Do you mean the `scramble/digest` block?

msfschaffner replied:
this description is out of date and will change once I've updated the block diagram...

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +296,418 @@ The following table lists the main parameters used throughout the OTP controller
+### Programming Already Programmed Regions
+
+OTP words cannot be programmed twice, and doing so may damage the memory array.
+Hence the OTP controller performs a blank check and returns an error if a write operation is issued to an already programmed location.
```
Is this error the "Programming error occurred" error you mentioned at line 450?

msfschaffner replied:
Yes, I this is one of the error conditions that could lead to 0x7. Depending on how many error conditions we can have during the programming step, we may have to split these up into multiple codes...

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +296,418 @@ The following table lists the main parameters used throughout the OTP controller
+
+Error Code | Description
+-----------|-------------
+0x0        | Command completed successfully.
+0x1        | Invalid command.
+0x2        | Read or write attempted before initialization.
+0x3        | Failure during OTP initialization.
+0x4        | Correctable read error occurred.
+0x5        | Uncorrectable read error occurred.
+0x6        | Other read errors (TBD).
+0x7        | Programming error occurred (TBD).
+others     | Reserved.
```
This err_code seems to be different from the ERR_CODE section in OTP register.

msfschaffner replied:
yeah they do not map one-to-one at the moment.

we have to finetune these error codes once the inner workings of the closed source wrapper are clearer. at the moment, these codes just serve as a placeholder. I'll add a TODO here mentioning this.

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
 
+The OTP is a module that provides a device with one-time-programming functionality.
+The result of this programming is non-volatile, and unlike flash, cannot be reversed.
+The OTP functionality is constructed through an open source OTP controller and a proprietary OTP IP.
```
jep, fixed.

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+- An open-source abstraction interface that software can use to interact with a proprietary OTP block underneath.
+- An open-source abstraction interface that hardware components (for example life cycle controller and key manager (**TODO: add links here**)) can use to interact with a proprietary OTP block underneath.
+- High level logical security protection, such as integrity checks and scrambling of sensitive content.
+- Software isolation for when OTP contents are readable and programmable
```
fixed

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,285 @@ title: "OTP Controller Technical Specification"
+Within each logical partition, there are specific enforceable properties
+
+- Secrecy
+  - This controls whether a particular partition contains secret data
+  - If secret, a partition is not readable by software once locked, and is encrypted in storage
```
jup, fixed.

msfschaffner commented:
Thanks for the reviews! I have addressed the comments in separate commits to ease subsequent reviews (these commits will be squashed once this PR is ready).

Also, I have revised the memory map and streamlined a couple of things, as discussed with @moidx, @tjaychen and Nuvoton in our latest call:
- all tokens are 128bit in size now
- removed individual locks for tokens and keys, and split the secure partition into 3 subpartitions with their own locks
- digests for 64bit partitions are 64bit in size, digests for 32bit partitions are 32bit in size
- updated life cycle state encoding
- some rewording of the affected sections

PTAL and let me know if that looks good to you.

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,290 @@ title: "OTP Controller Technical Specification"
+
+The functionality of OTP is split into an open-source and a closed-source part, with a clearly defined boundary in between, as illustrated in the simplified high-level block diagram below.
+
+![OTP Controller Overview](otp_controller_overview.svg)
```
image source here for reference:
https://docs.google.com/drawings/d/1SYxx3PSW54lIRkkJf3De0UMqlMC3y6lzBobitjOO7vw

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,290 @@ title: "OTP Controller Technical Specification"
+The "back-end" represents the programming interface used by hardware and software components to stage the upcoming values.
+The diagram below illustrates this behavioral model.
+
+![OTP Controller Block Diagram](otp_behavioral_model.svg)
```
image source here for reference:
https://docs.google.com/drawings/d/1QpMVnU3DSSNxm2agdC-895xGmfWJP4q34tsGCHVHxSw

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +301,461 @@ The following table lists the main parameters used throughout the OTP controller
+**TODO: revamp the diagram below and add more details**
+The following is a high-level block diagram that illustrates everything that has been discussed.
+
+![OTP Controller Block Diagram](otp_controller_block_diagram.svg)
```
image source here for reference (WIP):
https://docs.google.com/drawings/d/1nQ7fq3NLit5xozO9MUplAmi6wZ3r0ncHdw1xNZpDe1Y

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +301,461 @@ The following table lists the main parameters used throughout the OTP controller
+
+### Scrambling, Digest and Ephemeral Key Derivation Mechanism
+
+![OTP Digest Mechanism](otp_digest_mechanism.svg)
```
image source here for reference (WIP):
https://docs.google.com/drawings/d/17EEsLEzRIrfesEu3bXCWrCBFwk_Lw0Vz4CwOxU4Pjeo

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
image source here for reference:
https://docs.google.com/drawings/d/1fNXqDXGOR7iLRvLpSWunaQznHHbNEMubGmvAU7bwxPo

msfschaffner commented:
The latest push adds a detailed blockdiagram plus some further design detail explanations.
Apart from several small TODOs (like missing links), the main thing that remains now are the life-cycle and partition FSM details, which I will add as I work out the corresponding RTL blocks.

msfschaffner commented:
The latest push squashes all updates into a single commit.

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,290 @@ title: "OTP Controller Technical Specification"
+It also implies that no matter how the OTP storage or word size may change underneath, the open-source controller must present a consistent and coherent software and hardware interface.
+This standardized interface is defined further below, and the wrapper leverages the same [technology primitive mechanism]({{< relref "hw/ip/prim/README.md" >}}) that is employed in other parts of OpenTitan in order to wrap and abstract technology-specific macros (such as memories and clocking cells) that are potentially closed-source.
+
+In order to enable simulation and FPGA emulation of the OTP controller even without access to the proprietary OTP IP, a generalized and synthesizable model of the OTP IP is provided in the form of a generic technology primitive (`prim_generic_otp`).
```
maybe good to provide the path to prim_generic_top?

msfschaffner replied:
jep, let's do that.

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,290 @@ title: "OTP Controller Technical Specification"
+SECRET2        | yes    | yes      | yes (digest) | yes (digest) | RMA unlock token and creator root key.
+
+Generally speaking, the production life cycle of a device is split into 5 stages "Manufacturing" -> "Calibration and Testing" -> "Provisioning" -> "Mission" -> "RMA".
+OTP values are usually programmed during "Calibration and Testing", "Provisioning" and "RMA" stages, as explained below.
```
looks like "Calibration and Testing" and "Provisioning" have sessions below to talk to OTP behavior, do you think it is necessary to have a session for "RMA" as well?

cindychip commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -8,14 +8,290 @@ title: "OTP Controller Technical Specification"
+
+Once the appropriate partitions have been locked, the hardware integrity checker employs two integrity checks to verifies the content of the volatile buffer registers:
+
+1. All buffered partitions have additional Byte parity protection that is concurrently monitored.
```
Is there a reason `Byte` is capitalized?

msfschaffner replied:
Hm... I guess I do that out of habit since Byte is also capitalized when abbreviated (B), but that may not be entirely correct. let me fix those.

cindychip reviewed (commented):
Woa, nice doc and graphs. Thanks for the effort. Looks pretty good to me with a few small comments.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
do we need this check for `HW_CFG` and `SECRET`? Those will have digests right?

msfschaffner replied:
yes that is right. that is up for us to decide. it would certainly simplify the partition controllers a bit if we did that.

also, everything that is buffered (including LIFE_CYCLE) has Byte parity.
so, given that LIFE_CYCLE is redundantly encoded, you could argue that LIFE_CYCLE doesn't need this consistency check, neither...

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
is this part still true? Or are you thinking about doing this part differently?

msfschaffner replied:
this is still the plan yes. I did not want to compute the digest over the plaintext secrets, since the digests for the secret partitions remain readable by SW...

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
i wonder if it would be simpler on any error to never release...of course that would brick the device.

msfschaffner replied:
so implementation-wise I think it is simpler to handle this on a per-partition basis, since each partition has its own controller... 

note that these integrity/consistency checks are pretty severe and will anyways escalate to the alert handler...

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +301,495 @@ The following table lists the main parameters used throughout the OTP controller
+`otp_csrng_i`            | `input`          | `otp_csrng_rsp_t`           | Entropy acknowledgement from [CSRNG]({{< relref "hw/ip/csrng/doc" >}}).
+`pwr_otp_init_i`         | `input`          | `pwrmgr_pkg::pwr_otp_req_t` | Initialization request coming from power manager  **TODO: link to power manager docs**
+`pwr_otp_init_o`         | `output`         | `pwrmgr_pkg::pwr_otp_rsp_t` | Initialization response going to power manager  **TODO: link to power manager docs**
+`otp_pwr_state_o`        | `output`         | `otp_pwr_state_t`           | Current OTP state, used by power manager in order to determine whether a shutdown needs to be aborted due to an ongoing OTP operation.
```
Do you want to leave a note here that this MAY not be needed?

msfschaffner replied:
good idea - added a TODO for that.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +301,495 @@ The following table lists the main parameters used throughout the OTP controller
+`lc_otp_program_o`       | `output`         | `lc_otp_program_rsp_t`      | Life cycle state transition request. See also **TODO: link to life cycle docs**
+`lc_provision_en_i`      | `input`          | `lifecycle_pkg::lc_tx_t`    | Provision enable qualifier coming from life cycle controller. This signal enables read / write access to the RMA_TOKEN and CREATOR_ROOT_KEY_SHARE0 and CREATOR_ROOT_KEY_SHARE1. **TODO: link to life cycle docs**
+`lc_test_en_i`           | `input`          | `lifecycle_pkg::lc_tx_t`    | Test enable qualifier coming from from life cycle controller. This signals enables the TL-UL access port to the proprietary OTP IP. **TODO: link to life cycle docs**
+`otp_lc_data_o`          | `output`         | `otp_lc_data_t`             | Life cycle state output holding the current life cycle state and the value of the transition counter.
```
will this hold the tokens also?

msfschaffner replied:
yes, it basically holds everything the LC controller needs to know. let me add that.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
```diff
@@ -25,30 +301,495 @@ The following table lists the main parameters used throughout the OTP controller
+`lc_test_en_i`           | `input`          | `lifecycle_pkg::lc_tx_t`    | Test enable qualifier coming from from life cycle controller. This signals enables the TL-UL access port to the proprietary OTP IP. **TODO: link to life cycle docs**
+`otp_lc_data_o`          | `output`         | `otp_lc_data_t`             | Life cycle state output holding the current life cycle state and the value of the transition counter.
+`otp_keymgr_key_o`       | `output`         | `keymgr_key_t`              | Key output to the key manager holding CREATOR_ROOT_KEY_SHARE0 and CREATOR_ROOT_KEY_SHARE1.
+`otp_flash_key_o`        | `output`         | `keymgr_key_rsp_t`          | Key output to the key flash controller holding FLASH_DATA_KEY and FLASH_ADDR_KEY.
```
i can't completely remember now..but did you want a req/ack interface for the key consumers?

msfschaffner replied:
yeah that's the idea for the ephemeral scrambling keys, and the flash key yes. I have only transcribed here what I already have in the hjson, so the interface description for FLASH needs to be updated (I'll do that right now). 

for the CREATOR_ROOT_KEY I was thinking to just output the value once valid (i.e., something like KEY + valid signal), since that comes straight out of the partition and does not go through the scrambling datapath...

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
this sounds like a good idea.

msfschaffner replied:
ok, I'll take note and add a sentence describing this.

tjaychen commented on hw/ip/otp_ctrl/doc/_index.md:
maybe we could just source more bits from flash also? 512b is probably too much... but this might be a good opportunity to make it 256b or something and crunch down.

msfschaffner replied:
ah... so to clarify, this is already using 256bit to produce one 128bit key output. e.g., FLASH_DATA_KEY is 256bit, and each 128bit halve gets reduced into a 64bit halve. the two derived halves then make up the 128bit key that is actually used for flash scrambling...

do you have an area in flash where you could pack in more bits that we could feed into this?

msfschaffner replied:
(sorry I haven't updated the memory map to reflect this just yet)

msfschaffner replied:
ok I've adjusted the memory map to reflect this now.

tjaychen commented:
just reviewed the spec portion, it looks good to me.
I didn't see any info on the access_gate though.  I can't remember if that's meant to be the TAP access point from life cycle controller? Or was it from the normal debug module..?

tjaychen commented on hw/ip/otp_ctrl/data/otp_ctrl.hjson:
```diff
@@ -286,42 +298,69 @@
+            Register write enable for all direct access interface registers.
+            ''',
+      swaccess: "rw",
+      hwaccess: "hro",
```
do you think this should also be set-able by hardware? That way once you trigger a transaction, we prevent anything from changing.

msfschaffner replied:
ah that is a good point. let me do that, yes.

tjaychen commented on hw/ip/otp_ctrl/data/otp_ctrl.hjson:
```diff
@@ -452,165 +491,124 @@
+      }
+    },
+    { multireg: {
+        name:     "SECRET1_DIGEST",
```
is it still the case that this is calculated over the scrambled contents? Not the raw one right?

msfschaffner replied:
jep that is correct.

moidx reviewed (approved):
@msfschaffner Thanks!

LGTM as first draft. I added some minor comments. 

It will be good to use @cdgori's implementation guidance doc to review the encoding of the control signals.

Provisioning and key manager distribution LGTM.  I will start cross liking some of the programming model details from here into the provisioning specifications.

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
Sure I can rephrase this.

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
fixed

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
hm that is an interesting thought... would this be something like a device secret that we would program into a secret but unscrambled partition before provisioning the actual secrets?

msfschaffner replied:
@tjaychen

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
oh yeah, thanks

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
in all cases the digest serves as the write-lock "switch". the only difference in case of SW_CFG partitions is that it is up to SW to calculate, program and verify the digest.

msfschaffner replied:
@cdgori
@tjaychen

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
I think this is exactly the point. the values are buffered in plaintext form, hence a secure wipe via LFSR might be safer...

msfschaffner commented on hw/ip/otp_ctrl/doc/_index.md:
we intend to get some more numbers on that from the vendor in order to estimate this more accurately.

note however that the "integrity" check described here would only perform the check between data and digest buffered in registers. it is the purpose of the "consistency" check further below to crosscheck the buffered digests with the digests sitting in the OTP array. 

we are currently debating whether this consistency check is required at all, given the redundancy provided via the integrity checks. See https://github.com/lowRISC/opentitan/pull/3180#discussion_r48265356

Also, the consistency checks would increase the total number of read operations to be executed by OTP, which might have an impact on the total life time of the OTP array (need input from the vendor).

@cdgori 
@tjaychen 
@zi-v

msfschaffner commented:
Thanks for the review round, @moidx and @tjaychen. I've addressed most of the comments. Some aspects however still need to be further discussed before making additional changes (e.g. whether the consistency check is actually needed or not).
